File: backends/node-nestjs/src/validation.test.ts
import type { ArgumentMetadata } from "@nestjs/common";
import { describe, expect, it } from "vitest";
import { UserStatusBodyDto } from "./admin-users/user-status.dto.js";
import { BookmarkBodyDto } from "./bookmarks/bookmark.dto.js";
import { MessageBodyDto } from "./messages/message.dto.js";
import { BookmarkStatusBodyDto, ReportBodyDto, ResolutionBodyDto } from "./moderation/moderation.dto.js";
import { ValidationProblem } from "./problems.js";
import { contractValidationPipe } from "./validation.pipe.js";

const metadata = (metatype: ArgumentMetadata["metatype"]): ArgumentMetadata => ({
  type: "body",
  metatype,
});

async function transform<T extends object>(metatype: new () => T, body: unknown): Promise<T> {
  return contractValidationPipe().transform(body, metadata(metatype)) as Promise<T>;
}

async function violations<T extends object>(metatype: new () => T, body: unknown) {
  try {
    await transform(metatype, body);
    return [];
  } catch (error) {
    if (error instanceof ValidationProblem) return error.violations;
    throw error;
  }
}

describe("DTO validation metadata", () => {
  it("normalizes bookmark inputs and applies defaults", async () => {
    const input = await transform(BookmarkBodyDto, {
      url: " https://example.com ",
      title: " t ",
      tags: [" Node ", "node", "web"],
    });

    expect(input).toMatchObject({
      url: "https://example.com",
      title: "t",
      notes: null,
      tags: ["node", "web"],
      visibility: "private",
    });
  });

  it("maps malformed bookmark values to canonical keys without coercion", async () => {
    await expect(violations(BookmarkBodyDto, { title: "t" })).resolves.toContainEqual({
      field: "url",
      messageKey: "validation.url.required",
    });
    await expect(
      violations(BookmarkBodyDto, { url: "https://example.com", title: "t", tags: [42] }),
    ).resolves.toContainEqual({ field: "tags", messageKey: "validation.tag.invalid" });
    await expect(
      violations(BookmarkBodyDto, { url: "https://example.com", title: "t", notes: 42 }),
    ).resolves.toContainEqual({ field: "notes", messageKey: "validation.notes.too-long" });
  });

  it.each([
    [
      MessageBodyDto,
      { key: "example", language: "en", text: "x", description: 42 },
      "description",
      "validation.message.description.too-long",
    ],
    [ReportBodyDto, { reason: "spam", comment: 42 }, "comment", "validation.report.comment.too-long"],
    [ResolutionBodyDto, { resolution: "dismissed", note: 42 }, "note", "validation.resolution.note.too-long"],
    [BookmarkStatusBodyDto, { status: "active", note: 42 }, "note", "validation.bookmark-status.note.too-long"],
    [UserStatusBodyDto, { status: "active", reason: 42 }, "reason", "validation.block.reason.too-long"],
  ] as const)("rejects wrong optional scalar types for %s", async (metatype, body, field, messageKey) => {
    await expect(violations(metatype, body)).resolves.toContainEqual({ field, messageKey });
  });
});

File: backends/elixir-phoenix/test/validation_test.exs
defmodule StackverseBackend.ValidationTest do
  use ExUnit.Case, async: true

  alias StackverseBackend.Validation

  test "bookmark validation normalizes tags and defaults visibility" do
    assert {:ok, bookmark} =
             Validation.validate_bookmark(%{
               "url" => "https://example.com",
               "title" => " Example ",
               "tags" => [" Elixir ", "elixir", "phoenix"]
             })

    assert bookmark.title == "Example"
    assert bookmark.tags == ["elixir", "phoenix"]
    assert bookmark.visibility == "private"
  end

  test "bookmark validation collects field errors" do
    assert {:error, errors} =
             Validation.validate_bookmark(%{
               "url" => "ftp://example.com",
               "title" => "",
               "notes" => String.duplicate("x", 4001),
               "tags" => ["Bad Tag"]
             })

    keys = Enum.map(errors, & &1.message_key)
    assert "validation.url.invalid" in keys
    assert "validation.title.required" in keys
    assert "validation.notes.too-long" in keys
    assert "validation.tag.invalid" in keys
  end

  test "message validation enforces contract keys" do
    assert {:error, errors} =
             Validation.validate_message(%{
               "key" => "Bad Key",
               "language" => "english",
               "text" => "",
               "description" => String.duplicate("x", 1001)
             })

    keys = Enum.map(errors, & &1.message_key)
    assert "validation.message.key.invalid" in keys
    assert "validation.message.language.invalid" in keys
    assert "validation.message.text.required" in keys
    assert "validation.message.description.too-long" in keys
  end

  test "report validation accepts contract reasons only" do
    assert {:ok, %{reason: "spam"}} = Validation.validate_report(%{"reason" => "spam"})
    assert {:error, errors} = Validation.validate_report(%{"reason" => "unknown"})
    assert [%{message_key: "validation.report.reason.invalid"}] = errors
  end
end
